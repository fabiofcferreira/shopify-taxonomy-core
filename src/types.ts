import { z } from "zod";

export enum TaxonomyLanguage {
  Danish = "da",
  German = "de",
  English = "en",
  Spanish = "es",
  Finnish = "fi",
  French = "fr",
  Italian = "it",
  Japanese = "ja",
  Korean = "ko",
  Dutch = "nl",
  Polish = "pl",
  BrazilianPortuguese = "pt-BR",
  Portuguese = "pt-PT",
  Chinese = "zh-CN",
}

export const CATEGORY_ID_REGEXP = new RegExp(/^[a-z]{2}((-[0-9]+)*)$/);

const TaxonomyTreeVerticalCategoryValidator = z.object({
  id: z.string(),
  name: z.string(),
  parent_id: z.string().nullable(),
});

export type TaxonomyTreeVerticalCategory = z.infer<
  typeof TaxonomyTreeVerticalCategoryValidator
>;

export type TaxonomyCategory = TaxonomyTreeVerticalCategory & {
  fullName: string;
};

const TaxonomyTreeVerticalValidator = z.object({
  name: z.string(),
  prefix: z.string(),
  categories: TaxonomyTreeVerticalCategoryValidator.array(),
});

export type TaxonomyTreeVertical = z.infer<
  typeof TaxonomyTreeVerticalValidator
>;

export const TaxonomyTreeValidator = z.object({
  version: z.string(),
  verticals: TaxonomyTreeVerticalValidator.array(),
});

export type TaxonomyTree = z.infer<typeof TaxonomyTreeValidator>;
