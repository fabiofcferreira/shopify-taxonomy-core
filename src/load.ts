import { readFileSync } from "fs";
import path from "path";
import { TaxonomyLanguage, TaxonomyTreeValidator } from "./types";

export function loadTaxonomyTree(directory = "./", language: TaxonomyLanguage) {
  try {
    const taxonomyFilename = `${language}.json`;

    const fileBuffer = readFileSync(path.join(directory, taxonomyFilename));
    const fileContents = JSON.parse(Buffer.from(fileBuffer).toString());

    return TaxonomyTreeValidator.parse(fileContents);
  } catch (ex) {
    throw new Error(
      `Could not load taxonomy tree file: ${ex instanceof Error ? ex.message : ex}`,
    );
  }
}
