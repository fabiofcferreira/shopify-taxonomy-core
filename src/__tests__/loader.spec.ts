import path from "path";
import { TaxonomyLanguage } from "../types";
import { loadTaxonomyTree } from "../load";

const DEFINITIONS_FOLDER_PATH = path.join(__dirname, "definitions");

describe("Loader", () => {
  it("should load correctly when taxonomy definition file is found", () => {
    expect(() =>
      loadTaxonomyTree(DEFINITIONS_FOLDER_PATH, TaxonomyLanguage.English),
    ).not.toThrow();
  });

  it("should throw error when taxonomy definition file is not found", () => {
    expect(() =>
      loadTaxonomyTree(DEFINITIONS_FOLDER_PATH, TaxonomyLanguage.German),
    ).toThrowErrorMatchingSnapshot();
  });

  it("should throw error when taxonomy definition file schema is not unknown", () => {
    expect(() =>
      loadTaxonomyTree(DEFINITIONS_FOLDER_PATH, TaxonomyLanguage.Portuguese),
    ).toThrowErrorMatchingSnapshot();
  });
});
