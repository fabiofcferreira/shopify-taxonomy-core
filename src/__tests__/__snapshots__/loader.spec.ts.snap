// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader should throw error when taxonomy definition file is not found 1`] = `"Could not load taxonomy tree file: Error: ENOENT: no such file or directory, open '/Users/fabiofcferreira/code/personal/shopify-taxonomy-core/src/__tests__/definitions/de.json'"`;

exports[`Loader should throw error when taxonomy definition file schema is not unknown 1`] = `
"Could not load taxonomy tree file: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "version"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "verticals"
    ],
    "message": "Required"
  }
]"
`;
